libs/geo/src/main/java/org/elasticsearch/geometry/utils/SpatialEnvelopeVisitor.java

```diff
@@ -0,0 +1,302 @@
+import org.elasticsearch.geometry.Point;
+import org.elasticsearch.geometry.Polygon;
+import org.elasticsearch.geometry.Rectangle;
+
+import java.util.Optional;
+
+/**
+ * This visitor is designed to determine the spatial envelope (or BBOX or MBR) of a potentially complex geometry.
+ * It has two modes:
+ * <ul>
+ *     <li>
+ *         Cartesian mode: The envelope is determined by the minimum and maximum x/y coordinates.
+ *         Incoming BBOX geometries with minX &gt; maxX are treated as invalid.
+ *         Resulting BBOX geometries will always have minX &lt;= maxX.
+ *     </li>
+ *     <li>
+ *         Geographic mode: The envelope is determined by the minimum and maximum x/y coordinates,
+ *         considering the possibility of wrapping the longitude around the dateline.
+ *         A bounding box can be determined either by wrapping the longitude around the dateline or not,
+ *         and the smaller bounding box is chosen. It is possible to disable the wrapping of the longitude.
+ * </ul>
+ * Usage of this is as simple as:
+ * <code>
+ *     Optional&lt;Rectangle&gt; bbox = SpatialEnvelopeVisitor.visit(geometry);
+ *     if (bbox.isPresent()) {
+ *         Rectangle envelope = bbox.get();
+ *         // Do stuff with the envelope
+ *     }
+ * </code>
+ * It is also possible to create the inner <code>PointVisitor</code> separately, as well as use the visitor for multiple geometries.
+ * <code>
+ *     PointVisitor pointVisitor = new CartesianPointVisitor();
+ *     SpatialEnvelopeVisitor visitor = new SpatialEnvelopeVisitor(pointVisitor);
+ *     for (Geometry geometry : geometries) {
+ *         geometry.visit(visitor);
+ *     }
+ *     if (visitor.isValid()) {
+ *         Rectangle envelope = visitor.getResult();
+ *         // Do stuff with the envelope
+ *     }
+ * </code>
+ * Code that wishes to modify the behaviour of the visitor can implement the <code>PointVisitor</code> interface,
+ * or extend the existing implementations.
+ */
+public class SpatialEnvelopeVisitor implements GeometryVisitor<Boolean, RuntimeException> {
+
+    private final PointVisitor pointVisitor;
+
+    public SpatialEnvelopeVisitor(PointVisitor pointVisitor) {
+        this.pointVisitor = pointVisitor;
+    }
+
+    /**
+     * Determine the BBOX without considering the CRS or wrapping of the longitude.
+     * Note that incoming BBOX's that do cross the dateline (minx>maxx) will be treated as invalid.
+     */
+    public static Optional<Rectangle> visit(Geometry geometry) {
+        var visitor = new SpatialEnvelopeVisitor(new CartesianPointVisitor());
+        if (geometry.visit(visitor)) {
+            return Optional.of(visitor.getResult());
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * Determine the BBOX assuming the CRS is geographic (eg WGS84) and optionally wrapping the longitude around the dateline.
+     */
+    public static Optional<Rectangle> visit(Geometry geometry, boolean wrapLongitude) {
```

Comment:
1. Avoid overloading here: it isn't clear which `visit` refers to Cartesian and which to geo coordinates.
2. Replace `boolean` with an enum or an overload (between two geo shape versions) to avoid Boolean Blindness.

Reply:
I would use different names to specify what is geo and what is cartesian. no so fuzzy about the use of boolean here, enum is fine too.

Reply:
Used different names, by copying the version in ST_EXTENT PR. If we want to use an enum, we can make that change in that PR.